File: temp/SandBox/AStar.java
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.TreeSet;

public abstract class AStar<Node> {

  /**
   * Run A* starting from start node
   * If not found, return null and empty path
   * Else return goal and path contains trace back to start from goal
   */
  public final Node search(Node start, HashMap<Node, Node> path) {
    HashSet<Node> visited = new HashSet<Node>();
    final HashMap<Node, Double> score = new HashMap<Node, Double>();
    TreeSet<Node> queue = new TreeSet<Node>(new Comparator<Node>() {
      @Override
      public int compare(Node a, Node b) {
        return Double.compare(score.get(a) + heuristic(a), score.get(b) + heuristic(b));
      }
    });

    score.put(start, 0d);
    for(queue.add(start); !queue.isEmpty(); ) {
      Node c = queue.pollLast();
      if(isGoal(c)) {
        return c;
      }
      visited.add(c);
      for(Node n : neighbours(c)) {
        if (!visited.contains(n) && (!score.containsKey(n) || score.get(c) + distance(c, n) <= score.get(n))) {
          score.put(n, score.get(c) + distance(c, n));
          path.put(n, c);
          if (queue.contains(n)) {
            queue.remove(n);
          }
          queue.add(n);
        }
      }
    }

    path.clear();
    return null;
  }

  /**
   * @return all neighbours of node n
   */
  protected abstract Iterable<Node> neighbours(Node n);

  /**
   * Admissible heuristic distance from goal of node n
   * Must never over-estimate i.e. heuristic(x) <= dist(x,y) + heuristic(y) for all x,y
   * When h(x) = 0 then its a simple BFS
   */
  protected double heuristic(Node n) {
    return 0;
  }

  /**
   * returns true iff n is goal (by default we assume heuristic(n) is 0 iff n is goal)
   */
  protected boolean isGoal(Node n) {
    return heuristic(n) == 0;
  }

  /**
   * distance between two neighbour nodes
   * Note: This function will be called for all children
   */
  protected double distance(Node a, Node b) {
    return 1;
  }
}

File: src/main/scala/scalgos/AStar.scala
package scalgos

import collection.mutable

object AStar {

  /**
   * Run A* algorithm
   *
   * @param start starting node
   * @param isGoal true iff we are at goal
   * @param neighbors returns neighbors of given node
   * @param distance  distance between two neighbour nodes (guaranteed to be only called for @param neighbors)
   * @param heuristic admissible heuristic distance from goal to node n
   *                  must never over-estimate i.e. heuristic(x) <= dist(x,y) + heuristic(y) for all x,y
   *                  If not known, simply use 0
   * @tparam Node  encapsulates each position/state in search space
   * @return If found Some(goal, path) (where path.head is start and path.last is goal) else None
   */
  def run[Node](start: Node,
                isGoal: Node => Boolean,
                neighbors: Node => Iterable[Node],
                distance: (Node, Node) => Double = (i: Node, j: Node) => 1,
                heuristic: Node => Double = (i: Node) => 0d): Option[(Node, Seq[Node])] =
  {
    val score = mutable.Map(start -> 0d).withDefaultValue(Double.PositiveInfinity)
    val priority = Ordering by {n: Node => score(n) + heuristic(n)} reverse
    val queue = mutable.TreeSet(start)(priority)
    val parent = mutable.Map.empty[Node, Node]
    val visited = mutable.Set.empty[Node]

    def removeFirst = {
      val head = queue.head
      queue -= head
      head
    }

    while(!queue.isEmpty) {
      val c = removeFirst
      if (isGoal(c)) {
        val trace = mutable.ArrayBuffer.empty[Node]
        var current = c
        while (parent contains current) {
          current +=: trace
          current = parent(current)
        }
        assert(trace.head == start && trace.tail == c)
        return Some(c, trace.toSeq)
      }
      neighbors(c) filterNot visited.contains foreach {n =>
        if(score(n) >= score(c) + distance(c, n)) {
          queue -= n
          score(n) = score(c) + distance(c, n)
          queue += n
          parent(n) = c
        }
      }
      visited += c
    }
    None
  }
}
